crates/holochain_p2p/src/spawn/actor.rs

```diff
@@ -534,6 +543,157 @@ impl kitsune2_api::KitsuneHandler for HolochainP2pActor {
+#[derive(Debug)]
+struct BootWrap {
+    compat: NetworkCompatParams,
+    preflight: Arc<std::sync::Mutex<bytes::Bytes>>,
+    orig: kitsune2_api::DynBootstrap,
+    cache: std::sync::Mutex<Vec<Arc<AgentInfoSigned>>>,
+}
+
+impl kitsune2_api::Bootstrap for BootWrap {
+    fn put(&self, info: Arc<AgentInfoSigned>) {
+        let Self {
+            compat,
+            preflight,
+            orig,
+            cache,
+        } = self;
+
+        let agents = {
+            let mut cache = cache.lock().unwrap();
+
+            // remove expired infos and previous infos that match the
+            // one that was generated
+            let now = kitsune2_api::Timestamp::now();
+            cache.retain(|cache_info| {
+                if cache_info.expires_at < now {
+                    return false;
+                }
+                if cache_info.agent == info.agent && cache_info.space == info.space {
+                    return false;
+                }
+                true
+            });
+
+            // add the one that was generated
+            cache.push(info.clone());
+
+            let mut agents = Vec::new();
+
+            // string encoding
+            for agent in cache.iter() {
+                if let Ok(encoded) = agent.encode() {
+                    agents.push(encoded);
+                }
+            }
+
+            agents
+        };
+
+        // encode the preflight message and cache
+        if let Ok(encoded) = (crate::wire::WirePreflightMessage {
+            compat: compat.clone(),
+            agents,
+        })
+        .encode()
+        {
+            *preflight.lock().unwrap() = encoded;
+        }
+
+        // don't forget to invoke the original bootstrap handler
+        orig.put(info);
+    }
+}
+
+/// This factory wraps the original bootstrap factory, generating
+/// original bootstrap instances, and wrapping them with our wrapper.
+#[derive(Debug)]
+struct BootWrapFact {
+    compat: NetworkCompatParams,
+    preflight: Arc<std::sync::Mutex<bytes::Bytes>>,
+    orig: kitsune2_api::DynBootstrapFactory,
+}
+
+impl kitsune2_api::BootstrapFactory for BootWrapFact {
+    fn default_config(&self, config: &mut Config) -> K2Result<()> {
+        self.orig.default_config(config)
+    }
+
+    fn validate_config(&self, config: &Config) -> K2Result<()> {
+        self.orig.validate_config(config)
+    }
+
+    fn create(
+        &self,
+        builder: Arc<Builder>,
+        peer_store: DynPeerStore,
+        space: SpaceId,
+    ) -> BoxFut<'static, K2Result<DynBootstrap>> {
+        let compat = self.compat.clone();
+        let preflight = self.preflight.clone();
+        let orig_fut = self.orig.create(builder, peer_store, space);
+        Box::pin(async move {
+            let orig = orig_fut.await?;
+            let out: DynBootstrap = Arc::new(BootWrap {
+                compat,
+                preflight,
+                orig,
+                cache: std::sync::Mutex::new(Vec::new()),
```

Comment:
Does this have default?

```suggestion
                cache: Default::default(),
```